builtin/mainmenu/dlg_create_world.lua

```diff
@@ -51,46 +159,165 @@ local function create_world_formspec(dialogdata)
 				end
 			end
 		end
+
+		local ds = (gameconfig:get("disallowed_settings") or ""):split()
+		for _, value in pairs(ds) do
+			disallowed_settings[value:trim()] = true
+		end
 	end
 
 	local mglist = ""
-	local selindex = 1
+	local selindex
 	local i = 1
+	local first_mg
 	for k,v in pairs(mapgens) do
+		if not first_mg then
+			first_mg = v
+		end
 		if current_mg == v then
 			selindex = i
 		end
 		i = i + 1
 		mglist = mglist .. v .. ","
 	end
+	if not selindex then
+		selindex = 1
+		current_mg = first_mg
+	end
 	mglist = mglist:sub(1, -2)
 
+	local mg_main_flags = function(mapgen, y)
+		if mapgen == "singlenode" then
+			return "", y
+		end
+		if disallowed_settings["mg_flags"] then
+			return "", y
+		end
+
+		local form = "checkbox[4.2," .. y .. ";flag_mg_caves;" ..
+			fgettext("Caves") .. ";"..strflag(flags.main, "caves").."]"
+		y = y + 0.5
+
+		form = form .. "checkbox[4.2,"..y..";flag_mg_dungeons;" ..
+			fgettext("Dungeons") .. ";"..strflag(flags.main, "dungeons").."]"
+		y = y + 0.5
+
+		local d_name
+		if mapgen == "v6" then
+			d_name = fgettext("Decorations (without trees and jungle grass)")
```

Comment:
Thanks, this section looks simpler now.
This text might be better as something like:
"Decorations (no effect on trees and jungle grass)"
The current text is a little unclear.